pageserver/src/tenant_threads.rs

```diff
@@ -1,77 +1,173 @@
         error!("compact loop terminated with error: {:?}", err);
         Err(err)
     } else {
         Ok(())
     }
 }
 
-fn compact_loop_ext(tenantid: ZTenantId) -> Result<()> {
+async fn compaction_loop_ext(tenantid: ZTenantId) -> Result<()> {
     loop {
         if tenant_mgr::get_tenant_state(tenantid) != Some(TenantState::Active) {
             break;
         }
         let repo = tenant_mgr::get_repository_for_tenant(tenantid)?;
         let compaction_period = repo.get_compaction_period();
 
-        std::thread::sleep(compaction_period);
-        trace!("compaction thread for tenant {} waking up", tenantid);
+        tokio::time::sleep(compaction_period).await;
+        trace!("compaction loop for tenant {} waking up", tenantid);
 
         // Compact timelines
         let repo = tenant_mgr::get_repository_for_tenant(tenantid)?;
-        repo.compaction_iteration()?;
+        let compaction_result =
+            tokio::task::spawn_blocking(move || repo.compaction_iteration()).await;
+        match compaction_result {
+            Ok(Ok(())) => {
+                // success, do nothing
+            }
+            Ok(Err(e)) => {
+                anyhow::bail!(e.context("Compaction failed"));
+            }
+            Err(e) => {
+                anyhow::bail!("Compaction join error {}", e);
+            }
+        }
     }
 
     trace!(
-        "compaction thread stopped for tenant {} state is {:?}",
+        "compaction loop stopped for tenant {} state is {:?}",
         tenantid,
         tenant_mgr::get_tenant_state(tenantid)
     );
     Ok(())
 }
 
+static START_GC_LOOP: OnceCell<Sender<ZTenantId>> = OnceCell::new();
+static START_COMPACTION_LOOP: OnceCell<Sender<ZTenantId>> = OnceCell::new();
+
+/// Spawn a task that will periodically schedule garbage collection until
+/// the tenant becomes inactive. This should be called on tenant
+/// activation.
+pub fn start_gc_loop(tenantid: ZTenantId) -> Result<()> {
+    START_GC_LOOP
+        .get()
+        .unwrap()
+        .blocking_send(tenantid)
+        .unwrap();
+    Ok(())
+}
+
+/// Spawn a task that will periodically schedule compaction until
+/// the tenant becomes inactive. This should be called on tenant
+/// activation.
+pub fn start_compaction_loop(tenantid: ZTenantId) -> Result<()> {
+    START_COMPACTION_LOOP
+        .get()
+        .unwrap()
+        .blocking_send(tenantid)
+        .unwrap();
+    Ok(())
+}
+
+/// Spawn the TenantTaskManager
+/// This needs to be called before start_gc_loop or start_compaction_loop
+pub fn init_tenant_task_pool() -> Result<()> {
+    let runtime = tokio::runtime::Builder::new_multi_thread()
+        .thread_name("tenant-task-worker")
+        .worker_threads(40) // Way more than necessary
+        .max_blocking_threads(100) // Way more than necessary
+        .enable_all()
+        .build()?;
+
+    let (gc_send, mut gc_recv) = mpsc::channel::<ZTenantId>(100);
+    START_GC_LOOP.set(gc_send).unwrap();
```

Comment:
> -> Result<
> .unwrap(

One more semi-random agitation for better error handling. I also prefer `.expect(` in non-obvious places where we would want to panic for some reason, but even that is very close to `.context(..)?` already.

Reply:
I used `expect`, which is fine in this case. `.context(..)?` is a bit clunky on non-anyhow errors